File: elixir/112-PathSum/main.exs
defmodule TreeNode do
  @type t :: %__MODULE__{
          val: integer,
          left: TreeNode.t() | nil,
          right: TreeNode.t() | nil
        }
  defstruct val: 0, left: nil, right: nil
end

defmodule Solution do
  def has_path_sum(nil, _) do
    false
  end

  def has_path_sum(root, target_sum) when root.left == nil and root.right == nil do
    target_sum - root.val == 0
  end

  @spec has_path_sum(root :: TreeNode.t | nil, target_sum :: integer) :: boolean
  def has_path_sum(root, target_sum) do
    has_path_sum(root.left, target_sum - root.val) or has_path_sum(root.right, target_sum - root.val)
  end
end

defmodule Tests do
  def example1 do
    tree = %TreeNode{val: 1, left: %TreeNode{val: 2}, right: %TreeNode{val: 3}}
    IO.puts(Solution.has_path_sum(tree, 3))
  end

  def example2 do
    tree = %TreeNode{val: 1, left: %TreeNode{val: 2}, right: %TreeNode{val: 3}}
    IO.puts(Solution.has_path_sum(tree, 4))
  end

  def example3 do
    tree = %TreeNode{val: 1, left: %TreeNode{val: 2}, right: %TreeNode{val: 3}}
    IO.puts(Solution.has_path_sum(tree, 1))
  end
end

Tests.example1
Tests.example2
Tests.example3

File: python/112-PathSum/main.py
from typing import Optional

import sys
sys.path.append('../modules')
from leetcode import TreeNode

import unittest

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root is None:
            return False

        # We're at a leaf node
        if root.left is None and root.right is None and targetSum - root.val == 0:
            return True

        if root.left and self.hasPathSum(root.left, targetSum - root.val) :
            return True

        if root.right and self.hasPathSum(root.right, targetSum - root.val) :
            return True

        return False

    
class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        
        root = TreeNode.makeTree(1, 2, 3)
        self.assertTrue(solution.hasPathSum(root, 3))

    def test_example2(self):
        solution = Solution()
        
        root = TreeNode.makeTree(1, 2, 3)
        self.assertTrue(solution.hasPathSum(root, 4))

    def test_example3(self):
        solution = Solution()
        
        root = None
        self.assertFalse(solution.hasPathSum(root, 0))

    def test_example4(self):
        solution = Solution()
        
        root = TreeNode.makeTree(5, 4, 8, 11, None, 13, 4, 7, 2, None, None, None, 1)
        self.assertTrue(solution.hasPathSum(root, 22))


if __name__ == '__main__':
    unittest.main()
